Feature/l2 extension clis

sandip-dalvi opened the pull request:
Add new CLIs in Wi-Fi Shell L2 module.
1. wifi version - To get the driver/firmware version 
2. wifi ap bw - To get/set AP bandwidth configuration.

sandip-dalvi committed:
net: wifi: add wifi driver version API

Add command to query to WiFi driver/firmware revision. The API is expected
to return the firmware revision and driver version as a string, and can be
used by the user to determine what revision of the WiFi driver is in use.

Signed-off-by: Sandip Dalvi <sandip.dalvi@nxp.com>

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -53,6 +53,8 @@ extern "C" {
 enum net_request_wifi_cmd {
 	/** Get Wi-Fi driver and Firmware version */
 	NET_REQUEST_WIFI_CMD_VERSION = 1,
+	/** Set or get MAC Address */
```
This is already available as part of net-mgmt commands/APIs, so, no need to for a new one.

sandip-dalvi replied:
Didn't get your point

krish2718 replied:
Please see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/shell/iface.c#L238 `net iface <idx>` shows this info.

sandip-dalvi replied:
The API is for get/set both.

krish2718 replied:
We can do that too `set_mac`, MAC address is common across all network interfaces not just Wi-Fi, so, it belongs in the `net shell`.

```
$ net iface 1 help
iface: wrong parameter count
iface - Print information about network interfaces.
Subcommands:
  up       : 'net iface up <index>' takes network interface up.
  down     : 'net iface down <index>' takes network interface down.
  show     : 'net iface <index>' shows network interface information.
  set_mac  : 'net iface set_mac <index> <MAC>' sets MAC address for the network
            interface.
```

sandip-dalvi replied:
We can use that.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -51,8 +51,10 @@ extern "C" {
+	/** Get Wi-Fi driver and Firmware version */
+	NET_REQUEST_WIFI_CMD_VERSION = 1,
 	/** Scan for Wi-Fi networks */
-	NET_REQUEST_WIFI_CMD_SCAN = 1,
```
Do we really need to modify the existing enumeration? Can't we just add it at the end? I know version seems basic so, but it's ok to be at the last.

sandip-dalvi replied:
okay, I thought it is basic first command for any Wi-Fi related activity.

sandip-dalvi replied:
Updated

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -229,6 +236,14 @@ enum net_event_wifi_cmd {
+/** Wi-Fi version */
+struct wifi_version {
+	/** Driver version */
+	char *drv_version;
```
Who own's this memory? It's better to use an array with max version length configurable.

sandip-dalvi replied:
I agree. will update

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1838,6 +1838,28 @@ static int cmd_wifi_packet_filter(const struct shell *sh, size_t argc, char *arg
+		return -ENOEXEC;
+	}
+
+	struct wifi_version version = {0};
```
Should be at the begining.

sandip-dalvi replied:
Okay

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1886,6 +1908,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_twt_ops,
 );
 
 SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
+	SHELL_CMD(version, NULL, "Print Wi-Fi Driver and Firmware version",
```
`versions`

sandip-dalvi replied:
Updated

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -229,6 +236,14 @@ enum net_event_wifi_cmd {
 	(_NET_WIFI_EVENT | NET_EVENT_WIFI_CMD_AP_STA_DISCONNECTED)
 
+/** Wi-Fi version */
+struct wifi_version {
+	/** Driver version */
+	char drv_version[WIFI_DRV_VERSION_MAX_LEN];
+	/** Firmware version */
+	char fw_version[WIFI_FW_VERSION_MAX_LEN];
+};
```
Normally I would do it like this but this data structure is quite large and uses lot of stack. The data is probably mostly static and could be stored into ROM, and even if not so, the driver could use static char array to store the data only once. So I think we could safely use a pointer here as you originally had. Perhaps `const char * const drv_version` and similar for the `fw_version`

krish2718 replied:
Sure, but then we need to add checks to make sure point is still valid.

jukkar replied:
> Sure, but then we need to add checks to make sure point is still valid.

If the information is stored somewhere in ROM or in static array, then we can trust the pointer to be a valid one. It would be madness in the driver to store this kind of non changing data to a stack variable.

krish2718 replied:
Absolutely, but you never know :).

jukkar replied:
We can certainly document how the data pointed by the pointer should be stored. If the user wants to shoot itself to the foot by doing something weird, then go ahead and best luck :wink:

sandip-dalvi replied:
I agree with @jukkar to keep as pointers only, as these would be const ROM locations and will never change.

jukkar commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -83,6 +83,8 @@ enum net_request_wifi_cmd {
 	NET_REQUEST_WIFI_CMD_CHANNEL,
```
There should be no need for the mac setter/getter for wifi. As the wifi interface is built on top of ethernet interface, there is already MAC address setter function in `include/zephyr/net/ethernet_mgmt.h`
And the MAC address can be fetched by `net_if_get_link_addr()` function.
The net-shell has function to set the MAC address, but if you think it is missing some functionality, please propose a separate pr. The net-shell file is `subsys/net/lib/shell/iface.c`

sandip-dalvi replied:
We can use functionality from net-shell, Thanks for pointing out.

sandip-dalvi committed:
net: wifi: add wifi driver AP bandwidth API

Add command to query the AP bandwidth configuration. The API is expected
to return the current configured bandwidth or can be used by user to set
other valid bandwidth settings.

Signed-off-by: Sandip Dalvi <sandip.dalvi@nxp.com>
